lib/Optimizer/TrainingPreparation.cpp

```diff
@@ -0,0 +1,132 @@
+    if (BatchNormalizationNode::ScaleIdx == inputIdx) {
+      tensor->init(glow::Tensor::InitKind::Zero, 0, F->getPRNG());
+    } else if (BatchNormalizationNode::BiasIdx == inputIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    } else if (BatchNormalizationNode::MeanIdx == inputIdx) {
+      tensor->zero();
+    } else if (BatchNormalizationNode::VarIdx == inputIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 1.0, F->getPRNG());
+    }
+    break;
+  }
+  case Kinded::Kind::FullyConnectedNodeKind: {
+    if (FullyConnectedNode::WeightsIdx == inputIdx) {
+      FullyConnectedNode *FCN = llvm::cast<FullyConnectedNode>(node);
+      auto in = FCN->getInput();
+      tensor->init(Tensor::InitKind::Xavier, in.dims()[1], F->getPRNG());
+    } else if (FullyConnectedNode::BiasIdx == inputIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+    break;
+  }
+  case Kinded::Kind::SoftMaxNodeKind: {
+    if (SoftMaxNode::SelectedIdx == inputIdx) {
+      tensor->zero();
+    }
+    break;
+  }
+  default:
+    break;
+  }
+}
+} // namespace
+
+TensorInitializer getDefaultTensorInitializer() {
+  return defaultTensorInitializer;
+}
+
+llvm::Error prepareFunctionForTraining(Function *F,
+                                       PlaceholderBindings &bindings,
+                                       Placeholder *&selected,
+                                       TensorInitializer &&initializer) {
+
+  auto &nodes = F->getNodes();
+
+  selected = nullptr;
+  // Lookup all nodes, skip Storage types, enumerate inputs,
+  // replace Constant type with trainable Placeholders except special cases,
+  // like BatchNormalization inputs (mean and variance). In special cases
+  // replace Constant type with non-trainable Placeholders.
+  for (auto &node : nodes) {
+    // Skip storages.
+    if (llvm::isa<Storage>(&node)) {
+      continue;
+    }
+
+    const bool isSoftMax = node.getKind() == Kinded::Kind::SoftMaxNodeKind;
+    const bool isBatchNormalization =
+        node.getKind() == Kinded::Kind::BatchNormalizationNodeKind;
+
+    for (unsigned idx = 0, e = node.getNumInputs(); idx < e; ++idx) {
+      auto *IN = node.getNthInput(idx).getNode();
+      Constant *C = llvm::dyn_cast<Constant>(IN);
+      if (!C) {
+        continue;
+      }
+
+      // Condition for NON trainable case
+      // isSoftMax || isBatchNormalization &&
+      //  (BatchNormalizationNode::MeanIdx == idx ||
+      //   BatchNormalizationNode::VarIdx == idx)
+
+      const bool isTrainable =
+          !isSoftMax &&
+          (!isBatchNormalization || (BatchNormalizationNode::MeanIdx != idx &&
+                                     BatchNormalizationNode::VarIdx != idx));
+
+      auto *PH = F->getParent()->createPlaceholder(C->getType(), C->getName(),
+                                                   isTrainable);
+
+      if (isSoftMax) {
+        selected = PH;
```

Comment:
It's possible for there to be multiple SoftMax in a single Function, for example in the inception models where there are intermediate SoftMaxes. I think it makes more sense to remove this finding of the selected PH and just have the caller do it (just with a simple loop over the Function's Nodes) since it understands the design of the model better.

Reply:
The general problem even with this PR is that we assuming caller will have enough information about model topology. In practice, caller needs a generic API, which works out of the box for any model. The proposed API can be part of the other framework and should work as it is without additional work. I'm leaning to keep it. Anyway of "selected" is not found it will be nullptr and caller can take actions.

Reply:
The caller has to be aware at least of the inputs it wants to use for the model and associated names. We could always programmatically search through the Function for the corresponding input PHs based on their name to set their associated input Tensors. This is essentially what's going on when you do `loader.getNodeValueByName(inputName)`. You're just doing it differently for the SoftMax here because you know the model has a single SoftMax.

I think it's fine for now in this PR, but I think we should at least add a check here that selected is a nullptr before setting it, and if that fails say the pass only currently supports models with single SoftMax.